scripts/dts/gen_defines.py

```diff
@@ -146,34 +146,19 @@ def main():
         write_chosen(edt)
         write_global_compat_info(edt)
 
-        write_device_extern_header(args.device_header_out, edt)
+        write_device_ord_header(args.device_header_out, edt)
 
     if args.edt_pickle_out:
         write_pickled_edt(edt, args.edt_pickle_out)
 
 
-def write_device_extern_header(device_header_out, edt):
+def write_device_ord_header(device_ord_header_out, edt):
     # Generate header that will extern devicetree struct device's
 
-    with open(device_header_out, "w", encoding="utf-8") as dev_header_file:
-        print("#ifndef DEVICE_EXTERN_GEN_H", file=dev_header_file)
-        print("#define DEVICE_EXTERN_GEN_H", file=dev_header_file)
-        print("", file=dev_header_file)
-        print("#ifdef __cplusplus", file=dev_header_file)
-        print('extern "C" {', file=dev_header_file)
-        print("#endif", file=dev_header_file)
-        print("", file=dev_header_file)
-
+    with open(device_ord_header_out, "w", encoding="utf-8") as dev_ord_header_file:
         for node in sorted(edt.nodes, key=lambda node: node.dep_ordinal):
-            print(f"extern const struct device DEVICE_DT_NAME_GET(DT_{node.z_path_id}); /* dts_ord_{node.dep_ordinal} */",
-                  file=dev_header_file)
-
-        print("", file=dev_header_file)
-        print("#ifdef __cplusplus", file=dev_header_file)
-        print("}", file=dev_header_file)
-        print("#endif", file=dev_header_file)
-        print("", file=dev_header_file)
-        print("#endif /* DEVICE_EXTERN_GEN_H */", file=dev_header_file)
+            print(f"DEVICE_ORD_MACRO(DT_{node.z_path_id}); /* dts_ord_{node.dep_ordinal} */",
```

Comment:
This produces a non-self-contained header that takes DEVICE_ORD_MACRO as an implicit parameter. I'm familiar with the pattern, but I don't like it very much. It looks weird if you're unfamiliar with the code, and there are some tricky ways to use it incorrectly. The header is internal, so it's not terrible.

However, all other things being equal, I'd prefer it if the script produced a self-contained header, and the client files included it at the top and then called into something that it defined. Something like
```
#include "dts_ord.h"

#define DEVICE_MACRO(ord) \
    extern const struct device DEVICE_DT_NAME_GET(ord)

DT_ORD_FOREACH(DEVICE_MACRO);
```

I think in this scenario, you could have dts_ord.h look like this:
```
// Normal header include guards
#define DT_ORD_FOREACH(fn) \
    fn(DT_N_S_foo); \
    fn(DT_N_S_bar); \
    // ...
```

I think the Python script would be very close to how it is now with just the for loop body changed and an extra line before it.

That's a little bit weird, in that all the ordinal references end up in one giant macro, but the caller doesn't need to know that, and the `#include` and invocation look normal. To my mind, that would be an improvement.